### ISSUE TITLE
CustomConstraint not found / NaN values not supported in Inequality constraint?

### ISSUE THREAD

**Issue by nelsonrogers** [sdv-dev/SDV#1066]
### Environment Details

* SDV version: 0.17.1
* Python version: 3.7
* Operating System: MacOS

### Error Description

I have a simple multi table dataset (orders, products, customers). I was trying to test creating a custom constraint for the "orders" table that specified that the "order_date" had to be before the "shipped_date". However, when I try to fit the model, I get AttributeError: module 'sdv.constraints.tabular' has no attribute 'CustomConstraint'.

I tried to use the Inequality constraint but the "shipped_date" column has NaN values and I got the following error : 
```
---------------------------------------------------------------------------

MultipleConstraintsErrors                 Traceback (most recent call last)

[<ipython-input-170-89fae4524a02>](https://localhost:8080/#) in <module>
      3 model = HMA1(metadata)
      4 
----> 5 model.fit(tables)

8 frames

[/usr/local/lib/python3.7/dist-packages/sdv/metadata/table.py](https://localhost:8080/#) in _fit_constraints(self, data)
    440 
    441         if errors:
--> 442             raise MultipleConstraintsErrors('\n' + '\n\n'.join(map(str, errors)))
    443 
    444     def _transform_constraints(self, data, is_condition=False):

MultipleConstraintsErrors: 
Both high and low must be datetime.
```

### Steps to reproduce
Option 1:
I have incuded my notebook and CSV files

Option 2:
- Have a data table with two date columns (I have included the csv files I am using)
- Implement the following:
```
from sdv.constraints import create_custom_constraint

def is_valid(column_names, data):
        column_name_low=column_names[0]
        column_name_high=column_names[1]
    
        is_valid = (data[column_name2].isnull() or data[column_name_low] < data[column_name_high])

        return is_valid

DateInequality = create_custom_constraint(is_valid_fn=is_valid)

date_constraint = DateInequality(column_names=['order_date', 'shipped_date'], data=tables['orders'])
```
- add your table to the metadata with your new custom constraint
- Fit the model

This is the error I get : 
```
---------------------------------------------------------------------------

AttributeError                            Traceback (most recent call last)

[<ipython-input-154-89fae4524a02>](https://localhost:8080/#) in <module>
      3 model = HMA1(metadata)
      4 
----> 5 model.fit(tables)

10 frames

[/usr/local/lib/python3.7/dist-packages/sdv/constraints/base.py](https://localhost:8080/#) in import_object(obj)
     52     if isinstance(obj, str):
     53         package, name = obj.rsplit('.', 1)
---> 54         return getattr(importlib.import_module(package), name)
     55 
     56     return obj

AttributeError: module 'sdv.constraints.tabular' has no attribute 'CustomConstraint'
```

[customers.csv](https://github.com/sdv-dev/SDV/files/9785254/customers.csv)
[order_items.csv](https://github.com/sdv-dev/SDV/files/9785256/order_items.csv)
[order_statuses.csv](https://github.com/sdv-dev/SDV/files/9785257/order_statuses.csv)
[orders.csv](https://github.com/sdv-dev/SDV/files/9785258/orders.csv)
[products.csv](https://github.com/sdv-dev/SDV/files/9785259/products.csv)
[shippers.csv](https://github.com/sdv-dev/SDV/files/9785260/shippers.csv)
[Synthetic(1).ipynb.zip](https://github.com/sdv-dev/SDV/files/9785334/Synthetic.1.ipynb.zip)

**Comment by npatki** [comment]
Hi @nelsonrogers, nice to meet you. Really appreciate the detailed explanations, notebooks and data.

There may be a few different things going on -- let me start with some immediate issues that stand out to me:

1. The `Inquality` constraint is designed to account for missing values. The issue is that the involved columns (`order_date` and `shipped_date`) are marked as categorical when they really should be datetime. To fix this, I'd recommend converting the columns to datetimes (rather than leave them as strings):

```python
import pandas as pd

orders['order_date'] = pd.to_datetime(orders['order_date'])
orders['shipped_date'] = pd.to_datetime(orders['shipped_date'])
```

2. Does your database contain certain child rows (`orders`) that do not reference a parent row (`shippers`)?

```python
# can't have float as id (NaN values cause ints to become floats in this case)
orders['shipper_id'] = orders['shipper_id'].astype(float).astype('Int64')
```

In a multi-table case, the SDV expects that all child rows must be referencing a parent. Our algorithm doesn't account for orphan rows right now; we could make this into a new feature request.

As a workaround, would it be possible to modify the data in some way? Perhaps create a dummy `shipper` that represents `None` and use it as the parent.

Let me know if that makes sense. After fixing the above, let me know if that works.

**Comment by nelsonrogers** [comment]
Hi, thank you for the reply.

To answer the issues you raised:
1.  I tried converting the columns to `datetime` as you suggested. Unfortunately, the problem still remains - the `Inequality` constraint still seems to pose a problem (I also tried it once I had resolved the orphan irow issue). When I add the constraint
```
date_constraint = Inequality('order_date', 'shipped_date')
```
to the `orders` table, I  get the following error when calling the `sample` method:

```
---------------------------------------------------------------------------

IntCastingNaNError                        Traceback (most recent call last)

[<ipython-input-180-87345c6c0401>](https://localhost:8080/#) in <module>
      1 
----> 2 synthetic_data = model.sample()
      3 synthetic_data

22 frames

[/usr/local/lib/python3.7/dist-packages/pandas/core/dtypes/cast.py](https://localhost:8080/#) in astype_float_to_int_nansafe(values, dtype, copy)
   1212     if not np.isfinite(values).all():
   1213         raise IntCastingNaNError(
-> 1214             "Cannot convert non-finite values (NA or inf) to integer"
   1215         )
   1216     return values.astype(dtype, copy=copy)

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer
```

2. My database does indeed contain orphan rows. For example, if the order has not been shipped, there should be no shipper linked to the order. I have now created a dummy `shipper` which helps in generating data with no `shipper`. Thank you for the suggestion.

To summarize, the issues you raised have been corrected but the outcome remains the same (The `Inequality` constraint causes the error above, and the `CustomConstraint` returns the same error as indicated in my original comment)

**Comment by npatki** [comment]
Hi @nelsonrogers, I believe the `IntCastingNaNError` may be coming from the following line:

```python
# can't have float as id (NaN values cause ints to become floats in this case)
orders['shipper_id'] = orders['shipper_id'].astype(float).astype('Int64')
```

This the `Int64` data type doesn't appear to be compatible with the SDV software*. Is this line still part of your script? If so, I think it's safe to remove it since you have addressed the issues of orphan rows -- i.e. there shouldn't be any `NaN` values left in this column, so there's no need to convert it into ints.

*The `Int64` is a Python data type whereas the SDV library uses the `numpy` and `pandas` packages. If any conversations are needed, I'd suggest converting to the data types from these packages. For eg.

```python
import numpy as np

orders['shipper_id'] = orders['shipper_id'].astype(float).astype(np.int64)
```

**Comment by npatki** [comment]
Hi @nelsonrogers, were you able to fix the issue? I'll mark it as resolved since I think we found the underlying bug. 

Feel free to reply to this issue if you are still seeing any errors, and I will re-open the issue to investigate.

**Comment by nelsonrogers** [comment]
Hi @npatki, sorry for not getting back to you sooner.
Unfortunately the issue still remains despite making sure all dates are in the correct type and no longer converting to int64. The same errors still appear when calling model.sample().

I also still have the custom constraint issue.

**Comment by npatki** [comment]
Hi @nelsonrogers no problem. I was actually able to replicate the error when there are missing values in datetime columns that are involved in an `Inequality` constraint. Let's use https://github.com/sdv-dev/SDV/issues/1121 to focus on this bug and track a fix. 

> I assume this error could be due to trying to apply constraints across tables - I assumed it was supported but maybe not?

Constraints between 2 different tables are currently not supported -- but you are welcome to file a new feature request for this.

**Comment by npatki** [comment]
I think now we can close this issue as a duplicate of #1121? Happy to re-open if there's anything else!
